gflownet/utils/batch.py

```diff
@@ -0,0 +1,411 @@
+            self.parents_all = tfloat(
+                [p for parents in self.parents_all for p in parents],
+                device=self.device,
+                float_type=self.float,
+            )
+        elif self.loss == "trajectorybalance":
+            assert self.trajectory_indices is not None
+            parents_policy = torch.zeros_like(self.states_policy)
+            parents = torch.zeros_like(self.states)
+            for env_id, traj in self.trajectory_indices.items():
+                # parent is source
+                parents_policy[traj[0]] = tfloat(
+                    self.envs[env_id].state2policy(self.envs[env_id].source),
+                    device=self.device,
+                    float_type=self.float,
+                )
+                parents[traj[0]] = tfloat(
+                    self.envs[env_id].source,
+                    device=self.device,
+                    float_type=self.float,
+                )
+                # parent is not source
+                parents_policy[traj[1:]] = self.states_policy[traj[:-1]]
+                parents[traj[1:]] = self.states[traj[:-1]]
+            self.parents_policy = parents_policy
+            self.parents = parents
+
+    def merge(self, another_batch):
+        """
+        Merges two unprocessed batches.
+        """
+        if self.is_processed or another_batch.is_processed:
+            raise Exception("Cannot merge processed batches.")
+        if self.loss != another_batch.loss:
+            raise Exception("Cannot merge batches with different losses.")
+        self.envs.update(another_batch.envs)
+        self.states += another_batch.states
+        self.actions += another_batch.actions
+        self.done += another_batch.done
+        self.env_ids += another_batch.env_ids
+        self.masks_invalid_actions_forward += (
+            another_batch.masks_invalid_actions_forward
+        )
+        self.masks_invalid_actions_backward += (
+            another_batch.masks_invalid_actions_backward
+        )
+        self.parents += another_batch.parents
+        self.parents_all += another_batch.parents_all
+        self.parents_actions_all += another_batch.parents_actions_all
+        self.n_actions += another_batch.n_actions
+
+    def _process_trajectory_indices(self):
+        """
+        Obtains the indices in the batch that correspond to each env. It creates a dict
+        of trajectory indices (key: env_id, value: indices of the states in self.states
+        going in the order from s_1 to s_f. The dict is created and stored in the
+        self.trajectory_indices.
+        """
+        trajs = defaultdict(list)
+        for idx, (env_id, step) in enumerate(zip(self.env_ids, self.n_actions)):
+            trajs[env_id.item()].append((idx, step))
+        trajs = {
+            env_id: list(map(lambda x: x[0], sorted(traj, key=lambda x: x[1])))
+            for env_id, traj in trajs.items()
+        }
+        self.trajectory_indices = trajs
```

Comment:
Since we are making the `process_bla` methods return the relevant variables instead of silently change the attributes, I would also do it for this method.

Reply:
[Done](https://github.com/alexhernandezgarcia/gflownet/pull/111/commits/d09d740a51868c8c7022dd9dc7ec4b712b472f20)